configuration.org
#+TITLE: Emacs Configuration

* use-package initialization

#+BEGIN_SRC emacs-lisp
  (require 'package)

  (setq package-archives '(("org" . "https://orgmode.org/elpa/")
                           ("gnu" . "https://elpa.gnu.org/packages/")
                           ("melpa" . "https://melpa.org/packages/")))

  ;; initialize built-in package management
  (package-initialize)

  ;; update packages list if we are on a new install
  (unless package-archive-contents
    (package-refresh-contents))

  (require 'use-package)

  (setq my-packages '(use-package use-package-ensure gnu-elpa-keyring-update))
  (dolist (package my-packages)
    (unless (package-installed-p 'use-package)
      (package-install 'use-package)))

  (require 'use-package-ensure)
  (setq use-package-always-ensure t)

  (exec-path-from-shell-initialize)
#+END_SRC

* Global Settings
#+BEGIN_SRC emacs-lisp
(use-package smartparens
  :config (smartparens-global-mode 1))
(global-auto-revert-mode 1)
#+END_SRC

Smooth scrolling.
#+BEGIN_SRC emacs-lisp
(setq scroll-step 1
      scroll-conservatively 10000)
#+END_SRC

Bell
#+BEGIN_SRC emacs-lisp
(setq ring-bell-function 'ignore)
#+END_SRC

Mouse scrolling
#+BEGIN_SRC emacs-lisp
(setq mouse-wheel-scroll-amount '(1 ((shift) . 1) ((control) . nil)))
(setq mouse-wheel-progressive-speed nil)
#+END_SRC

Scrolling without moving cursor
#+BEGIN_SRC emacs-lisp
(global-set-key "\M-n" "\C-u1\C-v")
(global-set-key "\M-p" "\C-u1\M-v")
#+END_SRC

Truncating Long Lines
#+BEGIN_SRC emacs-lisp
(set-default 'truncate-lines t)
#+END_SRC

#+BEGIN_SRC emacs-lisp
(setq-default column-number-mode t)
#+END_SRC

Fill column stuff
#+BEGIN_SRC emacs-lisp
(use-package fill-column-indicator
  :config (setq-default auto-fill-mode t)
          (setq-default fill-column 80))
#+END_SRC

RevealJS
#+BEGIN_SRC emacs-lisp
  (use-package ox-reveal)
  (use-package htmlize)
#+END_SRC

Custom
#+BEGIN_SRC emacs-lisp
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(js-indent-level 8))
#+END_SRC

Org Mode
#+BEGIN_SRC emacs-lisp
  (use-package org)
#+END_SRC

Ace Jump Mode
#+BEGIN_SRC emacs-lisp
(use-package ace-jump-mode
 :bind ("C-." . ace-jump-mode))
#+END_SRC

Treemacs
#+begin_src emacs-lisp
  (use-package treemacs
    :config (progn
	      (treemacs-git-mode 'simple)
	      (treemacs-hide-gitignored-files-mode t))
    :bind ("M-0" . treemacs-select-window))

  (use-package treemacs-projectile
    :after (treemacs projectile))
#+end_src

Which-key
#+begin_src emacs-lisp
  (use-package which-key
    :config (which-key-mode))
#+end_src

Direnv
#+begin_src emacs-lisp
  (use-package direnv
    :config (direnv-mode))
#+end_src

Web Mode
#+BEGIN_SRC emacs-lisp
;;(setq web-mode-content-types-alist '(("jsx" . "\\.js[x]?\\'")))
  (add-hook 'web-mode-hook
        (lambda ()
          (message "Running new web-mode hook. web-mode-content-type: %s" web-mode-content-type)
          ;; short circuit js mode and just do everything in jsx-mode
          (if (equal web-mode-content-type "javascript")
              (web-mode-set-content-type "jsx")
            (message "now set to: %s" web-mode-content-type))))
#+END_SRC

Search Results: Use Same Buffer
#+begin_src emacs-lisp
  (add-to-list 'display-buffer-alist '("*grep" (display-buffer-reuse-window
                                                display-buffer-same-window)))
  (add-to-list 'display-buffer-alist '("*ripgrep" (display-buffer-reuse-window
                                                display-buffer-same-window)))
#+end_src

** Window Navigation
   #+begin_src emacs-lisp
     (windmove-default-keybindings)
   #+end_src

* Theming
#+BEGIN_SRC emacs-lisp
(load-theme 'tango-dark t)
#+END_SRC

* Language Settings
** All Programming Languages
#+BEGIN_SRC emacs-lisp
        (add-hook 'prog-mode-hook (lambda ()
                                    (subword-mode 1)
                                    (display-line-numbers-mode)))

        (use-package projectile
          :ensure t
          :init
          (setq projectile-project-search-path '("~/workspace/" "~/workspace/haskell"))
          (projectile-mode +1)

          :config
          (define-key projectile-mode-map (kbd "s-p") 'projectile-command-map)
          (define-key projectile-mode-map (kbd "C-c p") 'projectile-command-map)
          (setq projectile-completion-system 'ivy))

        (use-package projectile-ripgrep)

        (use-package paredit
          :hook ((clojure-mode scheme-mode lisp-mode lisp-interaction-mode emacs-lisp-mode org-mode) . paredit-mode))

  ;; (use-package yasnippet
  ;;   :config (yas-reload-all)
  ;;   :hook (prog-mode . (lambda ()
  ;;                        (subword-mode t)
  ;;                        (yas-minor-mode t)
  ;;                        (projectile-mode +1))))

        ;; Language Server Protocol (LSP)
        (use-package lsp-mode
          :init (progn
                  (setq lsp-headerline-breadcrumb-enable nil)
                  (setq lsp-keymap-prefix "C-c l"))
          ;; :hook (prog-mode . lsp-mode)
          :commands lsp
          ;;:bind (("M-." . lsp-find-definition))
          )
        ;;(require 'lsp-clients)
        (require 'lsp-ui)
        (add-hook 'lsp-mode-hook 'lsp-ui-mode)
        (add-hook 'lsp-mode-hook 'lsp-enable-imenu)

        (use-package lsp-ui
          :hook (prog-mode . lsp-ui-mode)
          ;; :config
          ;; (setq lsp-ui-doc-position 'bottom)
          )

        (use-package magit)

        (use-package dockerfile-mode
          :init (add-to-list 'auto-mode-alist '("Dockerfile\\'" . dockerfile-mode)))

        (use-package company
          :hook (prog-mode . company-mode))
#+END_SRC

** Binary
#+BEGIN_SRC emacs-lisp
;; hexl mode for binary files
(add-to-list 'auto-mode-alist '("\\.bin$" . hexl-mode))
#+END_SRC

** Clojure
#+BEGIN_SRC emacs-lisp
(use-package clojure-mode
  :hook (clojure-mode . (lambda ()
			  (show-paren-mode t)
			  (fci-mode 1)
			  (setq fci-rule-width 1)
			  (setq fci-rule-color "darkblue"))))
#+END_SRC

** CSS
#+BEGIN_SRC emacs-lisp
(add-hook 'css-mode-hook
  (lambda ()
    (setq css-indent-offset 2)
    (setq indent-tabs-mode nil)))
#+END_SRC

** Elm
   #+begin_src emacs-lisp
     (use-package elm-mode
       :config
       (setq indent-tabs-mode nil)
       (setq lsp-elm-elm-language-server-path "/home/tim/.nvm/versions/node/v16.4.0/bin/elm-language-server")
                                             ;;(setq elm-tags-on-save t)
       :hook (elm-mode . lsp)
       )
   #+end_src

** Graphviz
#+BEGIN_SRC emacs-lisp
(use-package graphviz-dot-mode)
#+END_SRC

** Haskell
#+BEGIN_SRC emacs-lisp
    (use-package haskell-mode
    :hook ((haskell-mode . direnv-mode))
       :bind (([f8] . haskell-navigate-imports)
              ("C-c C-l" . 'haskell-process-load-or-reload)
              ("C-c C-c" . 'haskell-compile)
              ("C-c C-z" . 'haskell-interactive-switch)
              ("C-c C-n C-t" . 'haskell-process-do-type)
              ("C-c C-n C-i" . 'haskell-process-do-info)
              ("C-c C-n C-c" . 'haskell-process-cabal-build)
              ("C-c C-n c" . 'haskell-process-cabal)
            ("M-." . 'xref-find-definitions))
      :custom
      (haskell-tags-on-save t)
      (haskell-stylish-on-save t))

    ;; (use-package lsp-haskell
    ;;   :defer t

    ;;   :init
    ;;   (add-hook 'haskell-mode-hook 'lsp-deferred)
    ;;   (add-hook 'haskell-literate-mode-hook #'lsp)

    ;;   :custom
    ;;   (lsp-haskell-plugin-import-lens-code-lens-on nil)
    ;;   (lsp-haskell-formatting-provider "stylish-haskell"))

    (font-lock-add-keywords
     'haskell-mode
     '(("-- \\(FIXME\\|TODO\\):" 1 font-lock-warning-face t)
       ;; undefined could be improved with a look ahead a la elisp
       ("[^(-- )][a-zA-Z]* \\(undefined\\)$" 1 font-lock-warning-face t)))

  (use-package flymake-flycheck)

  (use-package dante
    :ensure t
    :after haskell-mode
    :commands 'dante-mode
    :init
    (add-hook 'haskell-mode-hook 'flymake-mode)
    (remove-hook 'flymake-diagnostic-functions 'flymake-proc-legacy-flymake)
    (add-hook 'haskell-mode-hook 'dante-mode)
    (add-hook 'haskell-mode-hook
              (defun my-fix-hs-eldoc ()
                (setq eldoc-documentation-strategy #'eldoc-documentation-default)))
    :config
    (require 'flymake-flycheck)
    (defalias 'flymake-hlint
      (flymake-flycheck-diagnostic-function-for 'haskell-hlint))
    (add-to-list 'flymake-diagnostic-functions 'flymake-hlint))

#+END_SRC

** HTML
#+BEGIN_SRC emacs-lisp
(setq html-helper-build-new-buffer t)
#+END_SRC

** JasmineJS
** Java
#+BEGIN_SRC emacs-lisp
  (use-package lsp-java
    ;; :hook (java-mode . lsp)
    :config (add-hook 'java-mode-hook 'lsp))
#+END_SRC
** Markdown
#+BEGIN_SRC emacs-lisp
(use-package markdown-mode
  :init (custom-set-variables '(markdown-command "pandoc"))
  :hook (markdown-mode . (lambda ()
                           (fci-mode t)
                           (auto-fill-mode t))))
#+END_SRC

#+BEGIN_SRC emacs-lisp
(add-hook 'jasminejs-mode-hook (lambda () (jasminejs-add-snippets-to-yas-snippet-dirs)))
#+END_SRC

** Nix
   #+begin_src emacs-lisp
     (use-package nix-mode)
   #+end_src

** Octave
#+BEGIN_SRC emacs-lisp
(add-to-list 'auto-mode-alist '("\\.m$" . octave-mode))
#+END_SRC

** PHP
#+BEGIN_SRC emacs-lisp
(use-package php-mode
  :hook (php-mode . (lambda ()
                        (setq c-basic-offset 2)
			(c-set-offset 'case-label '+) ;; used for properly indenting switch statements.
			;;(lsp)
			)))
#+END_SRC

** Purescript
#+BEGIN_SRC emacs-lisp
  (use-package purescript-mode
    :hook (purescript-mode . turn-on-purescript-indent))
#+END_SRC

** Python
#+BEGIN_SRC emacs-lisp
(use-package lsp-python-ms
  :hook (python-mode . (lambda ()
                         (require 'lsp-python-ms)
                         (setq indent-tabs-mode t)
                         (setq python-indent-offset 4)
                         (setq tab-width 4)
                         ;;(setq c-basic-offset 2)
                         ;;(c-set-offset 'case-label '+) ;; used for properly indenting switch statements.
                         (lsp))))
(setq lsp-python-ms-executable
      "~/python-language-server/output/bin/Release/linux-x64/publish/Microsoft.Python.LanguageServer")
#+END_SRC

#+BEGIN_SRC emacs-lisp
  (use-package ttl-mode
    :mode "\\.ttl\\'")
#+END_SRC

** Scala
#+BEGIN_SRC emacs-lisp
;;(require 'scala-mode2)
(add-to-list 'auto-mode-alist '("\\.scala$" . scala-mode))
(add-hook 'scala-mode-hook '(lambda ()
			      (setq indent-tabs-mode nil)))
#+END_SRC

** SPARQL
#+begin_src elisp
  (use-package sparql-mode
    :mode "\\.sparql\\'")
#+end_src
** Typescript
#+BEGIN_SRC emacs-lisp
;; (defun setup-tide-mode ()
;;   (interactive)
;;   (tide-setup)
;;   (flycheck-mode +1)
;;   (setq flycheck-check-syntax-automatically '(save mode-enabled))
;;   (eldoc-mode +1)
;;   (tide-hl-identifier-mode +1)
;;   (company-mode +1)
;;   (lsp))
;;(add-hook 'typescript-mode-hook #'setup-tide-mode)

;; Add nvm directory to `exec-path` so that lsp-mode can find the
;; Javascript/Typescript language server (javascript-typescript-stdio)
(setq exec-path (append exec-path '("/home/tim/.nvm/versions/node/v11.1.0/bin")))
(setenv "PATH" (concat "/home/tim/.nvm/versions/node/v11.1.0/bin:" (getenv "PATH")))

(setq lsp-clients-typescript-server "typescript-language-server")
(setq lsp-clients-typescript-server-args '("--stdio"))

(add-hook 'typescript-mode-hook 'lsp)
#+END_SRC

** YAML
#+BEGIN_SRC emacs-lisp
(use-package yaml-mode)
#+END_SRC

* Miscellaneous
** Multiple Cursors
   #+begin_src emacs-lisp
     (use-package multiple-cursors
       :bind ("C->" . mc/mark-next-like-this))
   #+end_src


configuration.el

(require 'package)
(add-to-list 'package-archives '("marmalade" . "http://marmalade-repo.org/packages/"))
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/") t)
(add-to-list 'package-archives '("org" . "http://orgmode.org/elpa/") t)
(package-initialize)

(global-auto-revert-mode 1)
(smartparens-global-mode 1)

(setq scroll-step 1
      scroll-conservatively 10000)

(setq mouse-wheel-scroll-amount '(1 ((shift) . 1) ((control) . nil)))
(setq mouse-wheel-progressive-speed nil)

(require 'auto-complete)
(require 'auto-complete-config)
(global-auto-complete-mode t)
(ac-config-default)

(setq-default column-number-mode t)

(require 'fill-column-indicator)
(setq-default auto-fill-mode t)
(setq-default fill-column 80)

(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(js-indent-level 8))

;; (require 'org-export-as-s5)

(load-theme 'tango-dark t)

(require 'yasnippet)
(yas-reload-all)
(add-hook 'prog-mode-hook (lambda ()
                            (linum-mode)
                            (subword-mode)
                            (global-flycheck-mode)
                            (yas-minor-mode)))

(add-hook 'text-mode-hook 'turn-on-visual-line-mode)

(require 'clojure-mode)
(add-hook 'clojure-mode-hook
          '(lambda ()
             (paredit-mode t)
             (show-paren-mode t)
             (fci-mode 1)
             (setq fci-rule-width 1)
             (setq fci-rule-color "darkblue")))

(add-hook 'emacs-lisp-mode-hook
          (lambda ()
            (paredit-mode)))

(add-hook 'nrepl-mode-hook
          (lambda ()
            (paredit-mode)))

;; hexl mode for binary files
(add-to-list 'auto-mode-alist '("\\.bin$" . hexl-mode))

(add-to-list 'auto-mode-alist '("\\.m$" . octave-mode))

(add-hook 'php-mode-hook '(lambda ()
                            (setq c-basic-offset 2)))
  ;;   (require 'web-mode)
  ;;   (require 'yasnippet)
  ;;   (yas-global-mode 1)
;; (require 'php-mode)
  ;;   (require 'php-auto-yasnippets)
  ;;   (payas/ac-setup)
  ;;   (define-key php-mode-map (kbd "C-c C-y") 'yas/create-php-snippet)
  ;;   (add-to-list 'auto-mode-alist '("\\.php$" . web-mode))
  ;;   (add-hook 'web-mode-hook
  ;;          '(lambda ()
  ;;             ;; (smartparens-mode)
  ;;             ;; (setq web-mode-ac-sources-alist
  ;;             ;;            '(("php" . (ac-source-yasnippet ac-source-php-auto-yasnippets))
  ;;             ;;              ("html" . (ac-source-emmet-html-aliases ac-source-emmet-html-snippets))
  ;;             ;;              ("css" . (ac-source-css-property ac-source-emmet-css-snippets))))
  ;;             (setq web-mode-ac-sources-alist
  ;;                   '(("css" . (ac-source-words-in-buffer ac-source-css-property))
  ;;                     ("html" . (ac-source-words-in-buffer ac-source-abbrev))
  ;;                     ("php" . (ac-source-words-in-buffer
  ;;                               ac-source-words-in-same-mode-buffers
  ;;                               ac-source-dictionary
  ;;                               ac-source-php-auto-yasnippets
  ;;                               ac-source-yasnippet))))))

(setq html-helper-build-new-buffer t)

(add-hook 'haskell-mode-hook 'turn-on-haskell-indentation)
(add-hook 'haskell-cabal-mode-hook '(lambda ()
                                      (setq indent-tabs-mode nil)))

(font-lock-add-keywords
 'haskell-mode
 '(("-- \\(FIXME\\|TODO\\):" 1 font-lock-warning-face t)
   ;; undefined could be improved with a look ahead a la elisp
   ("[^(-- )][a-zA-Z]* \\(undefined\\)$" 1 font-lock-warning-face t)))

;;(require 'scala-mode2)
(add-to-list 'auto-mode-alist '("\\.scala$" . scala-mode))
(add-hook 'scala-mode-hook '(lambda ()
                              paredit-mode
                              (setq indent-tabs-mode nil)))

(setq refheap-user "tmciver")
(setq refheap-token "3501c201-ad54-4b7f-9f55-c41b83a297a1")
(put 'downcase-region 'disabled nil)
(custom-set-variables '(markdown-command "pandoc"))
